### PR TITLE
Add support for bytes types in tsp

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
This uncovered bad codegen for slices of base-64 encoded data. The helper for marshaling byte arrays now properly handles slices of base-64 encode data (i.e. [][]byte).
Track when to generate SerDe helpers instead of using heuristics which tend to miss corner cases.

### CODE REVIEW

File: packages/typespec-go/test/cadlranch/encode/bytesgroup/custom_client_test.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+func TestRequestBodyClientBase64url(t *testing.T) {
+	client, err := bytesgroup.NewRequestBodyClient(nil)
+	require.NoError(t, err)
+	resp, err := client.Base64url(context.Background(), []byte("test"), nil)
+	require.NoError(t, err)
+	require.Zero(t, resp)
+}
+
+func TestRequestBodyClientCustomContentType(t *testing.T) {
+	t.Skip("https://github.com/Azure/typespec-azure/issues/15")
```
Comment (jhendrixMSFT) [inline packages/typespec-go/test/cadlranch/encode/bytesgroup/custom_client_test.go:142]: This is preventing us from generating binary request/response payloads.

**Review by chlowell — APPROVED** [review]

File: packages/typespec-go/test/cadlranch/encode/bytesgroup/zz_header_client.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	urlPath := "/encode/bytes/header/base64"
+	req, err := runtime.NewRequest(ctx, http.MethodGet, runtime.JoinPaths(host, urlPath))
+	if err != nil {
+		return nil, err
+	}
+	req.Raw().Header["value"] = []string{base64.StdEncoding.EncodeToString(value)}
+	return req, nil
+}
+
+func (client *HeaderClient) Base64url(ctx context.Context, value []byte, options *HeaderClientBase64urlOptions) (HeaderClientBase64urlResponse, error) {
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply packages/typespec-go/test/cadlranch/encode/bytesgroup/zz_header_client.go:54]: Good catch! Will fix up the naming in a separate PR.
